### DISCUSSION TITLE
Shapefile polygons to pyvista

### DISCUSSION THREAD

**Discussion by RichardScottOZ** [pyvista/pyvista#2325] [Q&A]
I know there were a couple of old things around I have seen in the past.

What would be the suggested fast method currently - e.g. if have a large shapefile?

**Comment by RichardScottOZ** [comment]
Things of interest 

https://hatarilabs.com/ih-en/tutorial-to-convert-geospatial-data-shapefile-to-3d-data-vtk-with-python-geopandas-pyvista [geopandas]
https://github.com/OpenGeoVis/PVGeo-Examples/blob/master/2.1%20-%20Loading-Shapefiles-To-VTK.ipynb @banesullivan  [pyshp and vtk]
https://github.com/paulo-herrera/PyGTV - which I haven't tried - has a gui

https://stackoverflow.com/questions/66863007/how-to-extrude-an-irregular-polygon @adeak answer

**Comment by RichardScottOZ** [comment]
Could rasterise etc - but was wondering about methods - e.g. take a bunch of units, extend to depth [extrude?] and build proto-3d-model here.

**Comment by RichardScottOZ** [comment]
and polygon ordering?

**Comment by RichardScottOZ** [comment]
```python
rng = np.random.default_rng()

# create dummy data
N = 10
angles = np.linspace(0, 2*np.pi, N, endpoint=False)
radii = rng.uniform(0.5, 1.5, N)
coords = np.array([np.cos(angles), np.sin(angles)]) * radii
points_2d = coords.T  # shape (N, 2)

# embed in 3d, create polygon
# embed in 3d, create filled polygon
points_3d = np.pad(points_2d, [(0, 0), (0, 1)])  # shape (N, 3)
face = [N + 1] + list(range(N)) + [0]  # cell connectivity for a single cell
polygon = pv.PolyData(points_3d, faces=face)

polygon.plot()
# extrude along z and plot
p2 = polygon.extrude((0, 0, 0.5))
p2.plot()
```
![image](https://user-images.githubusercontent.com/72196131/158000399-7fab3aea-cd6d-46b3-b565-7ccf5403303a.png)
![image](https://user-images.githubusercontent.com/72196131/158000402-9f291840-301f-4f2f-af7b-8ff1f63e7953.png)

```python
p2 = polygon.extrude((0, 0, 0.5), capped=True)
p2.plot()
```
```
points_3d = np.pad(points_2d, [(0, 0), (0, 1)])  # shape (N, 3)
face = [N + 1] + list(range(N)) + [0]  # cell connectivity for a single cell
polygon = pv.PolyData(points_3d, faces=face)
# extrude along z and plot
body = polygon.extrude((0, 0, 0.5), capping=True)
body.plot(color='white', specular=1)
# extrude along z and plot
```
![image](https://user-images.githubusercontent.com/72196131/158000505-52c59053-8c07-4fbf-8ba9-95f45f57f230.png)

@adeak - capping is good

**Reply by RichardScottOZ** [reply]
![image](https://user-images.githubusercontent.com/72196131/158000588-e5457ec0-746c-4753-b82b-a4e34790d1dd.png)

- two more cells = the top and bottom faces?

**Reply by adeak** [reply]
Yup.

**Comment by whophil** [comment]
I'm not totally familiar with the shapefile format, but I recently ran into this issue when trying to convert Shapely polygons to VTK data structures. Polygons with no "holes" are easy enough - a polygon can be represented by a single cell in VTK.

For polygons with holes, such as the one below, I couldn't figure out a solution which did not involve a constrained Delaunay triangulation. Further, I couldn't get VTK's constrained delaunay to work consistently, especially for thin-shelled polygons, so I ended up using https://pypi.org/project/triangle/ for my application. [This thread](https://pyvista.slack.com/archives/CBA5095UH/p1644372096830669) has more issues on the Delaunay problems.
<img src=https://user-images.githubusercontent.com/2186528/158207809-6cdd6bc9-e180-42c1-a94b-268b0db19bae.png width=200px>

**Comment by RichardScottOZ** [comment]
Thanks very much Phil.
